### ISSUE TITLE
Proxy Support

### ISSUE THREAD

**Issue by douglaszickuhr** [StevenMMortimer/salesforcer#32]
Hey Steven

Thanks for your hard work on this package. It's just great and very appreciated.

When I am using httr I normally have to set the proxy manually (using proxy_settings function from httr) due to my company network particularities. Is that a way to set the proxy definition within salesforcer?

Thanks
Douglas

**Comment by StevenMMortimer** [comment]
@douglaszickuhr I'm not very familiar with `use_proxy()` in **httr**, but from the documentation it looks like you need to set it for every HTTP call (e.g. GET, POST, PUT, PATCH, DELETE). Is that correct? If so, I think I would have to create a pass-through option for additional httr arguments that all functions in the **salesforcer** package would accept. 

Could you try working with the code below to help me get a working example? This example should pull back a list of objects in your Salesforce org using the REST API.

```
library(httr)
library(salesforcer)
sf_auth()

list_objects_url <- sprintf("%s/services/data/v%s/%s",
                            salesforcer_state()$instance_url,
                            getOption("salesforcer.api_version"), 
                            "sobjects/")
httr_response <- GET(list_objects_url, use_proxy("64.251.21.73", 8080))
response_parsed <- content(httr_response, encoding="UTF-8")
response_parsed
```

Let me know if you have any other thoughts or feedback on how to implement.

**Comment by douglaszickuhr** [comment]
Hey Steven, thanks for your reply. 

Yes that's right I wrongly mentioned a function `proxy_settings()` but that is actually a wrapper function I built myself. There's a small issue with the code you provided as the `sf_auth()` process won't work because it also requires the proxy definition for the token request step.

The following reprex should work fine if you have the credentials + client details to authenticate with Salesforce.

I've added an additional argument to my `use_proxy()`as my company's network uses NTLM Proxy authentication, which you probably won't be able to test. I can't use `use_proxy("64.251.21.73", 8080)` here as well.

``` r
library(httr)
library(dplyr)

POST(url = "https://login.salesforce.com/services/oauth2/token",
     body = list(grant_type = "password",
                 client_id = "client_id",
                 clent_secret = "client_secret",
                 username = "username",
                 password = "pass"),
     encode = 'form',
     use_proxy(url = "ironportproxy-2",
               username = "",
               password = "",
               port = 8080,
               auth = "ntlm")) %>%
    content(encoding="UTF-8")
#> $error
#> [1] "invalid_client_id"
#> 
#> $error_description
#> [1] "client identifier invalid"
```

<sup>Created on 2019-08-12 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by douglaszickuhr** [comment]
Hey Steven, hope you're good. I've been working on a workaround. Would you mind to give your thoughts about?

I've created a new function called `my_proxy()` which basically is a wrapper for `use_proxy()` from `httr`. It checks some System Environments through `Sys.getenv()` function  and returns an httr's request object.

```r
my_proxy <- function(){
  if (Sys.getenv("proxy_url") != ""){
    use_proxy(url = Sys.getenv("proxy_url"),
              port = as.numeric(Sys.getenv("proxy_port")),
              username = Sys.getenv("proxy_username"),
              password = Sys.getenv("proxy_password"),
              auth = Sys.getenv("proxy_auth")
    )
  } else {
    NULL
  }
}
```
That function then might be calleed on every `out` assignment inside `VERB_n()`.
```r
VERB_n <- function(VERB, n = 5) {
  function(url, headers=character(0), ...) {
    
    current_state <- salesforcer_state()
    for (i in seq_len(n)) {
      
      if(is.null(current_state$auth_method)){
        out <- VERB(url = url, add_headers(headers), my_proxy(), ...)
      } else if(current_state$auth_method == 'OAuth'){
        if(grepl("/services/data/v[0-9]{2}.[0-9]{1}/jobs/ingest", url)){
          out <- VERB(url = url, add_headers(c(headers, "Authorization"=sprintf("Bearer %s", current_state$token$credentials$access_token))), my_proxy(), ...)  
        } else if(grepl("/services/async", url)){
          out <- VERB(url = url, add_headers(c(headers, "X-SFDC-Session"=current_state$token$credentials$access_token)), my_proxy(), ...)  
        } else {
          out <- VERB(url = url, config=config(token=current_state$token), add_headers(headers), my_proxy(), The...)  
        }
      } else if (current_state$auth_method == 'Basic') {
        if(grepl("/services/async", url)){
          out <- VERB(url = url, add_headers(c(headers, "X-SFDC-Session"=current_state$session_id)), my_proxy(), ...)
        } else {
          out <- VERB(url = url, add_headers(c(headers, "Authorization"=sprintf("Bearer %s", current_state$session_id))), my_proxy(), ...)  
        }
      }
      
      status <- status_code(out)
      if (status < 500 || i == n) break
      backoff <- runif(n = 1, min = 0, max = 2 ^ i - 1)
      ## TO DO: honor a verbose argument or option
      mess <- paste("HTTP error %s on attempt %d ...\n",
                    "  backing off %0.2f seconds, retrying")
      mpf(mess, status, i, backoff)
      Sys.sleep(backoff)
    }
    out
  }
}
```
I wondered why you have implemented that using a for loop and it looks like it's like a retrying process, isn't it? It might be significantly simplified by using the function [`RETRY()`](https://github.com/r-lib/httr/blob/master/R/retry.R) from `httr` as well. What do you think?

There's also the `sf_auth()` which requires a special kinda configuration definition but can still use `my_proxy()`.

```r
if (is.null(token)) {
      
      sf_oauth_app <- oauth_app("salesforce",
                                key = consumer_key, 
                                secret = consumer_secret,
                                redirect_uri = callback_url)
      
      sf_oauth_endpoints <- oauth_endpoint(request = NULL,
                                           base_url = sprintf("%s/services/oauth2", login_url),
                                           authorize = "authorize", access = "token", revoke = "revoke")
      
      sf_token <- oauth2.0_token(endpoint = sf_oauth_endpoints,
                                 app = sf_oauth_app, 
                                 cache = cache,
                                 config_init = my_proxy())
  
      stopifnot(is_legit_token(sf_token, verbose = TRUE))
      
      # set the global .state variable
      .state$auth_method <- "OAuth"
      .state$token <- sf_token
      .state$session_id <- NULL
      .state$instance_url <- sf_token$credentials$instance_url
      
    }
```

It has worked quite well here on my company network. 

```r
library(salesforcer)

Sys.setenv("proxy_url" = "ironportproxy-2")
Sys.setenv("proxy_port" = 8080)
Sys.setenv("proxy_auth" = "ntlm")

sf_auth()
#>Waiting for authentication in browser...
#>Press Esc/Ctrl + C to abort
#>Authentication complete.

sf_user_info()
#> $`accessibilityMode`
#>[1] "false"
#>
#>$chatterExternal
#>[1] "false"
#>
#>$currencySymbol
#>[1] NA
#>
#>$orgAttachmentFileSizeLimit
#>[1] "5242880"
#>
#>$orgDefaultCurrencyIsoCode
#>[1] NA
#>
#>$orgDefaultCurrencyLocale
#>[1] NA
#>
#>$orgDisallowHtmlAttachments
#>[1] "false"
#>
#>$orgHasPersonAccounts
#>[1] "true"
#>
#>$organizationId
#>[1] "[redacted]"
#>
#>$organizationMultiCurrency
#>[1] "true"
#>
#>$organizationName
#>[1] "[redacted]"
#>
#>$profileId
#>[1] "[redacted]"
#>
#>$roleId
#>[1] "[redacted]"
#>
#>$sessionSecondsValid
#>[1] "7200"
#>
#>$userDefaultCurrencyIsoCode
#>[1] "EUR"
#>
#>$userEmail
#>[1] "[redacted]"
#>
#>$userFullName
#>[1] "Douglas Zickuhr"
#>
#>$userId
#>[1] "[redacted]"
#>
#>$userLanguage
#>[1] "en_US"
#>
#>$userLocale
#>[1] "en_IE"
#>
#>$userName
#>[1] "[redacted]"
#>
#>$userTimeZone
#>[1] "Europe/Dublin"
#>
#>$userType
#>[1] "Standard"
#>
#>$userUiSkin
#>[1] "Theme3
```

**Comment by StevenMMortimer** [comment]
@douglaszickuhr Thanks! This looks good. I'll look into how to plug it into everything with the package. And thanks for the tip about `RETRY()` that didn't exist when I started the package.

**Comment by douglaszickuhr** [comment]
Hey @StevenMMortimer No problem! I hope it helps you somehow. `RETRY()` is super handy and basically abstracts that loop. 
I can't stress enough how much I appreciate your hard work on `salesforcer`. Thank you very much! The package is just great. :100:

**Comment by StevenMMortimer** [comment]
@douglaszickuhr It took me way to long to get around to this issue, but better late than never! Thanks for your patience and thanks for really digging into the code to show me how you implemented the proxy. I've pretty much replicated that and upgraded to using the `RETRY()` function as you pointed out. 

It should be as simple as setting the proxy arguments as a package option like this: 

```r
library(salesforcer)

options(salesforcer.proxy_url = "ironportproxy-2")
options(salesforcer.proxy_port = 8080)
options(salesforcer.proxy_username = NULL) # replace as needed!
options(salesforcer.proxy_password = NULL) # replace as needed!
options(salesforcer.proxy_auth = "ntlm")

# log in
sf_auth()

# print user info to check connectivity
sf_user_info()
```

This is now included in the latest release of the package (0.1.4) that is now on CRAN. Please let me know if you have problems and we can work through them. If everything works great or I don't hear back in a couple weeks, then I'll close this issue. Note, v0.1.4 includes other updates that are noted in [NEWS.md](https://github.com/StevenMMortimer/salesforcer/blob/master/NEWS.md)), as well.

**Thanks again for your assistance and your patience!** 👍
